### PR TITLE
feat: build against SPDK 21.01

### PR DESCRIPTION
**Pull request by jonathan-teh**
Fixes CAS-674

### CODE REVIEW

**Comment by jonathan-teh** [comment]
bors try

File: mayastor/src/bdev/nexus/nexus_bdev.rs [hunk 1]
```diff
@@ -643,6 +643,12 @@ impl Nexus {
 
         match errno_result_from_i32((), errno) {
             Ok(_) => {
+                // spdk_bdev_register registers a poller to wait for the
+                // examine to complete before calling spdk_notify_send with a
+                // pointer to the bdev name, which could be gone if the bdev
+                // was destroyed too soon after creation.
+                // Poll a few times to have that run.
+                Reactors::current().poll_times(3);
```
Comment (Hoverbear) [inline mayastor/src/bdev/nexus/nexus_bdev.rs:651]: Did you pick this by fair dice roll?

Comment (jonathan-teh) [reply mayastor/src/bdev/nexus/nexus_bdev.rs:651]: Actually I iteratively increased the number until the core test passed :) You could contrive another test that'll make it fail again. I think this'll do as a temporary band-aid until an SPDK-only reproducer can be created and then reported upstream.

Comment (mtzaurus) [reply mayastor/src/bdev/nexus/nexus_bdev.rs:651]: This code will go away once we switch to our controller-based devices.

**Review by mtzaurus — APPROVED** [review]

File: mayastor/src/bdev/nexus/nexus_bdev.rs (hunk 1 above)
Comment (tiagolobocastro) [reply mayastor/src/bdev/nexus/nexus_bdev.rs:651]: How likely is this to be an issue in kubernetes? and would it be worth increasing the number of polls to get some wiggle room?

Comment (jonathan-teh) [reply mayastor/src/bdev/nexus/nexus_bdev.rs:651]: This affects very short-lived bdevs so as long they're not being created and destroyed in quick succession it should be fine.

**Review by jkryl — APPROVED** [review]
Cool! thanks for bringing us up to the latest released spdk ❤️

**Comment by jonathan-teh** [comment]
bors merge

File: mayastor/src/bdev/nexus/nexus_bdev.rs (hunk 1 above)
Comment (jonathan-teh) [reply mayastor/src/bdev/nexus/nexus_bdev.rs:651]: Apparently 3 might not be enough as the latest CI build still failed the `core_4` test. I'm going to increase it to 10.

**Comment by jonathan-teh** [comment]
bors merge

**Comment by jonathan-teh** [comment]
bors merge

**Comment by jonathan-teh** [comment]
bors try

**Review by cjones1024 — APPROVED** [review]

**Comment by jonathan-teh** [comment]
I've removed the dodgy polling-after-bdev-register in favour of proper fix our openebs/spdk fork. However, I'm getting intermittent hangs (unrelated to that fix) in the `core_5` test after it closes the nexus child so I'm not merging this just yet.

**Comment by jonathan-teh** [comment]
bors try

**Comment by jonathan-teh** [comment]
I think the intermittent `core_5` hangs are due to changes in openebs/spdk@590b6e75 ("nbd: Use async manner to stop nbd device.").

**Comment by jonathan-teh** [comment]
> I think the intermittent `core_5` hangs are due to changes in [openebs/spdk@590b6e7](https://github.com/openebs/spdk/commit/590b6e75) ("nbd: Use async manner to stop nbd device.").

That turned out to be a red herring, it appears the culprit is openebs/spdk@d2704a0 ("lib/nbd: Fix the nbd_cleanup_io logic") which seems to be an attempt to fix the nbd device hanging. It removes early returns for hard disconnect in `nbd_io_exec()` and `nbd_io_xmit()`.

**Comment by jonathan-teh** [comment]
bors cancel

**Comment by jonathan-teh** [comment]
Got distracted and forgot to post my previous comment, I'm preparing a commit to revert that SPDK commit.

**Comment by jonathan-teh** [comment]
I've added 2 changes to our SPDK fork which are openebs/spdk@6392568 and openebs/spdk@3716462.

**Comment by jonathan-teh** [comment]
bors try

**Comment by gila** [comment]
bors merge

**Comment by gila** [comment]
borse merge

**Comment by jonathan-teh** [comment]
bors merge

**Pull request merged**